process/message.go

```diff
@@ -0,0 +1,404 @@
+		propose.round,
+		propose.block,
+		propose.validRound,
+	})
+}
+
+// UnmarshalJSON implements the `json.Unmarshaler` interface for the Propose type.
+func (propose *Propose) UnmarshalJSON(data []byte) error {
+	tmp := struct {
+		Sig        id.Signature `json:"sig"`
+		Signatory  id.Signatory `json:"signatory"`
+		Height     block.Height `json:"height"`
+		Round      block.Round  `json:"round"`
+		Block      block.Block  `json:"block"`
+		ValidRound block.Round  `json:"validRound"`
+	}{}
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+	propose.sig = tmp.Sig
+	propose.signatory = tmp.Signatory
+	propose.height = tmp.Height
+	propose.round = tmp.Round
+	propose.block = tmp.Block
+	propose.validRound = tmp.ValidRound
+	return nil
+}
+
+type Prevotes []Prevote
+
+type Prevote struct {
+	signatory id.Signatory
+	sig       id.Signature
+	height    block.Height
+	round     block.Round
+	blockHash id.Hash
+}
+
+func NewPrevote(height block.Height, round block.Round, blockHash id.Hash) *Prevote {
+	return &Prevote{
+		height:    height,
+		round:     round,
+		blockHash: blockHash,
+	}
+}
+
+func (prevote *Prevote) Signatory() id.Signatory {
+	return prevote.signatory
+}
+
+func (prevote *Prevote) SigHash() id.Hash {
+	return sha3.Sum256([]byte(fmt.Sprintf("Prevote(Height=%v,Round=%v,BlockHash=%v)", prevote.Height(), prevote.Round(), prevote.BlockHash())))
+}
+
+func (prevote *Prevote) Sig() id.Signature {
+	return prevote.sig
+}
+
+func (prevote *Prevote) Height() block.Height {
+	return prevote.height
+}
+
+func (prevote *Prevote) Round() block.Round {
+	return prevote.round
+}
+
+func (prevote *Prevote) BlockHash() id.Hash {
+	return prevote.blockHash
+}
+
+// MarshalJSON implements the `json.Marshaler` interface for the Prevote type.
+func (prevote Prevote) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		Sig       id.Signature `json:"sig"`
+		Signatory id.Signatory `json:"signatory"`
+		Height    block.Height `json:"height"`
+		Round     block.Round  `json:"round"`
+		BlockHash id.Hash      `json:"blockHash"`
+	}{
+		prevote.sig,
+		prevote.signatory,
+		prevote.height,
+		prevote.round,
+		prevote.blockHash,
+	})
+}
+
+// UnmarshalJSON implements the `json.Unmarshaler` interface for the Prevote type.
+func (prevote *Prevote) UnmarshalJSON(data []byte) error {
+	tmp := struct {
+		Sig       id.Signature `json:"sig"`
+		Signatory id.Signatory `json:"signatory"`
+		Height    block.Height `json:"height"`
+		Round     block.Round  `json:"round"`
+		BlockHash id.Hash      `json:"blockHash"`
+	}{}
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+	prevote.sig = tmp.Sig
+	prevote.signatory = tmp.Signatory
+	prevote.height = tmp.Height
+	prevote.round = tmp.Round
+	prevote.blockHash = tmp.BlockHash
+	return nil
+}
+
+type Precommits []Precommit
+
+type Precommit struct {
+	signatory id.Signatory
+	sig       id.Signature
+	height    block.Height
+	round     block.Round
+	blockHash id.Hash
```

Comment:
This is implicit in the algorithm, since there are `2f+1` honest replicas that will not broadcast a `Precommit` unless they have seen `2f+1 Prevotes`.

Reply:
@ross-pure would you agree with that?

Reply:
Yes, but what if dishonest replicas send out a `Precommit` without seeing a `Prevote`?

Reply:
`2f+1` honest replicas will not do this. That's why they're honest.

Reply:
My understanding was that the algorithm is designed to work with a `PREVOTE` only containing `id(v)`. Having a look at the proof of correctness I'm fairly sure that this understanding is correct - there shouldn't be a need to include the `PREVOTE`s in a `PRECOMMIT`.

Reply:
This is also my understanding. Resolving.